File: Assets/_Scripts/States/Puck/PuckLooseState.cs
using RageCure.StateUtils;
using Styx.Entities;
using Styx.Entities.PuckModule;
using Styx.Managers;
using UnityEngine;

using PlayerMessage = Styx.Entities.PlayerModule.Message;
using PuckMessage = Styx.Entities.PuckModule.Message;

namespace Styx.States
{
    public class PuckLooseState : State<PuckState>
    {
        private PuckState _puckState;
        public PuckState PuckState { get
            {
                if (_puckState == null)
                {
                    _puckState = PuckManager.Instance.GetPuckState();
                }
                return _puckState;
            }
            private set {
                _puckState = value;
            }
        }

        public override void Enter(PuckState puckState)
        {}

        public override void Execute(PuckState puckState)
        {}

        public override void Exit(PuckState puckState)
        {}

        public override void FixedExecute(PuckState puckState)
        {}

        public void OnTriggerEnter(Collider other)
        {
            if (!PuckState.StateMachine.IsInState(this))
                return;

            //TODO: Find a way to not use hardcoded string to reference the layer name
            if (other.gameObject.layer == LayerMask.NameToLayer("Player"))
            {
                var playerId = other.gameObject.GetComponent<PlayerId>().Value;
                var playerState = PlayerManager.Instance.GetPlayerState(playerId);
                playerState.Reduce(PlayerMessage.UpdateToCarryPuck);

                PuckState.Reduce(PuckMessage.UpdateToCarried);
            }
        }
    }
}

File: Assembly-FSharp/Assembly-FSharp/States/Puck/PuckLooseState.fs
namespace Styx.States.Puck

open UnityEngine

open RageCure.StateUtils
open Styx.Entities.PuckModule
open Styx.Managers
open Styx.Entities.PlayerModule
open Styx.Entities


type PuckLooseState () =
    inherit MonoBehaviour ()

    let updateToPuckState (state: PlayerState)  =
        PlayerModule.reduce PlayerModule.Message.UpdateToCarryPuck state

    let updateToCarried (state: PuckState) =
        PuckModule.reduce PuckModule.Message.UpdateToCarried state

    interface IState<Puck> with

        override this.Enter (entity: Puck) =
            ()
        override this.Execute (entity: Puck) =
            ()
        override this.FixedExecute (entity: Puck) =
            ()
        override this.Exit (entity: Puck) =
            ()

    member this.OnTriggerEnter(other: Collider) =
        if other.gameObject.layer = LayerMask.NameToLayer "Player" then
            PuckManager.Instance.GetPuck()
            |> updateToCarried
            |> focusPuck
            |> PuckManager.Instance.UpdatePuck

            other.gameObject.GetComponent<PlayerId>().Value
            |> PlayerManager.Instance.GetPlayerState
            |> updateToPuckState
            |> focusPlayer
            |> PlayerManager.Instance.UpdatePlayer 
            |> ignore
            
            

